lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,809 @@
+      if (inputLIDs.contains(argIdx)) {
+        nextLID = inputLIDs[argIdx];
+      }
+    } else {
+      nextLID = getOrCreateOpLID(next);
+    }
+
+    // Sanity check
+    assert(nextLID != noLID);
+
+    // Generate the ite for the register update reset condition
+    // i.e. reg <= reset ? 0 : next
+    genIte(next.getDefiningOp(), resetLID, constToLIDMap.at({0, width}),
+           nextLID, width);
+
+    // Finally generate the next statement
+    genNext(next, reg, width);
+  }
+
+public:
+  /// Visitor Methods used later on for pattern matching
+
+  // Visitor for the inputs of the module.
+  // This will generate additional sorts and input declaration explicitly for
+  // btor2 Note that outputs are ignored in btor2 as they do not contribute to
+  // the final assertions
+  void visit(hw::PortInfo &port) {
+    // Separate the inputs from outputs and generate the first btor2 lines for
+    // input declaration We only consider ports with an explicit bit-width (so
+    // ignore clocks)
+    if (port.isInput() && !port.type.isa<seq::ClockType>()) {
+      // Generate the associated btor declaration for the inputs
+      StringRef iName = port.getName();
+
+      // Guarantees that a sort will exist for the generation of this port's
+      // translation into btor2
+      int64_t w = requireSort(port.type);
+
+      // Record the defining operation's line ID (the module itself in the case
+      // of ports)
+      inputLIDs[port.argNum] = lid;
+
+      // We assume that the explicit name is always %reset for reset ports
+      if (iName == resetStr)
+        resetLID = lid;
+
+      genInput(w, iName);
+    }
+  }
+
+  // Emits the associated btor2 operation for a constant. Note that for
+  // simplicity, we will only emit `constd` in order to avoid bit-string
+  // conversions
+  void visitTypeOp(hw::ConstantOp op) {
+    // Make sure that a sort has been created for our operation
+    int64_t w = requireSort(op.getType());
+
+    // Prepare for for const generation by extracting the const value and
+    // generting the btor2 string
+    int64_t value = op.getValue().getSExtValue();
+    genConst(value, w, op);
+  }
+
+  // Wires can generally be ignored in bto2, however we do need
+  // to keep track of the new alias it creates
+  void visit(hw::WireOp op) {
+    // Retrieve the aliased operation
+    Operation *defOp = op.getOperand().getDefiningOp();
+    // Wires don't output anything so just record alias
+    setOpAlias(op, defOp);
+  }
+
+  void visitTypeOp(Operation *op) { visitInvalidTypeOp(op); }
+
+  // Handles non-hw operations
+  void visitInvalidTypeOp(Operation *op) {
+    // Try comb ops
+    dispatchCombinationalVisitor(op);
+  }
+
+  // Binary operations are all emitted the same way, so we can group them into
+  // a single method.
+  void visitBinOp(Operation *op, StringRef inst) {
+    TypeSwitch<Operation *, void>(op)
+        .template Case<
+            // All supported binary ops
```

Comment:
Is this necessary? It looks like this is only called from the visitors which already declaratively say which ops are supported.

Reply:
it's just a faster way to cast it down into the correct type. It's the same debate of many overloads vs. a big typeswitch. I personally thing that in this case a big type switch is what I need.

Reply:
Ah yes I see, that makes sense, I'll update it to remove the typeswitch